### DISCUSSION TITLE
Lowering latency when using Python API

### DISCUSSION THREAD

**Discussion by navarrastar** [streamlink/streamlink#5713] [Q&A] [answered]
### Description

My goal is to get the most recent as possible frame from a twitch stream and run it through my ML model that makes predictions on it. When I use the streamlink CLI to get the stream to play in vlc with `streamlink twitch.tv/ninja 480p --twitch-low-latency --twitch-disable-ads` I am very happy with how little delay there is. However, when I use the Python API to get the stream and read the frames with cv2, there is considerably more delay (2-3 seconds). My question is, how would you recommend I go about getting the delay down to the same amount as it is in vlc, while also being able to read the frames so I can send them through my model?

This is my current script (without my ML code):
```
import cv2
from streamlink import Streamlink

session = Streamlink()

session.set_option("hls-live-edge", 1)
session.set_option("twitch-low-latency", True)
session.set_option("hls-segment-stream-data", True)
session.set_option("twitch-disable-ads", True)

streams = session.streams("twitch.tv/ninja")
url = streams["480p"].url

cap = cv2.VideoCapture(url)

while True:

    success, frame = cap.read()
    cv2.imshow("stream", frame)
    if cv2.waitKey(1) & 0xFF == ord("q"):
        break

```
I not sure if the `session.set_option("twitch-low-latency", True)` and `session.set_option("twitch-disable-ads", True)` lines are actually doing anything because those are not in the documentation, but I figured I might as well try them and they don't throw an error.

If cv2 is what is adding the delay, do you have any recommendations for what library I should use instead? I've looked into ffmpeg, gstreamer, and using cv2 with cuda, but I want to be sure that cv2 is the bottleneck before I go diving into these.

Last thing, I do not really care about reading consecutive frames. Ideally, I would just get the most recent frame on a specified interval (every 0.5 seconds for example) and not worry about the other frames. In other words, the stream does not need to be "watchable" by a person and frequent buffering is okay, my goal is to capture the the most recently available frame every so often. I'm not saying that I don't want to hear possible solutions that involve me reading every frame, i'm just giving more context to my problem.

I appreciate any and all responses and suggestions, thank you.

**Comment by bastimeyer** [comment]
> I not sure if the `session.set_option("twitch-low-latency", True)` and `session.set_option("twitch-disable-ads", True)` lines are actually doing anything

They do not.

See the `option` keyword on the `Plugin` constructor:
- https://streamlink.github.io/api/plugin.html#streamlink.plugin.Plugin
- https://streamlink.github.io/api/options.html#streamlink.options.Options

Or see the migration page of the old/removed `Session.get_plugin_option()` and `Session.set_plugin_option()` methods:
- https://streamlink.github.io/migrations.html#streamlink-g-s-et-plugin-option

**Reply by bastimeyer** [reply, marked as answer]
But according to your code example, you are not using Streamlink at all, so what you're asking is completely irrelevant. You are only using Streamlink as a stream URL resolver by accessing the `url` attribute. You are not using Streamlink's HLS implementation.

**Reply by navarrastar** [reply]
oops, thanks for your help!
